File: kevin/travel-agent/bench.py
# -*- coding: utf-8 -*-
"""性能与 token 基准：冷启动 vs 缓存命中、串行 vs 并行执行对比。python bench.py

输出三组数据，用于验收时量化"加速响应 / 减少 token"两项优化：
  1. 冷启动全链路耗时（含真实天气 API + 知识库检索与缓存写入）
  2. 缓存命中全链路耗时（0 网络、0 token）
  3. 工具串行 vs 并行微基准（DAG 并行加速的依据）
"""
from __future__ import annotations

import sys
import time
from concurrent.futures import ThreadPoolExecutor

if hasattr(sys.stdout, "reconfigure"):
    try:
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    except Exception:
        pass

import tools  # noqa: F401 触发工具注册
from core.agent import TravelAgent
from core.llm import TOKEN_METER
from tools.base import invalidate

GOAL = "从上海出发去苏州2天，2人，预算3000元"


def meter_snapshot():
    return {"calls": TOKEN_METER.calls,
            "tokens": TOKEN_METER.prompt_tokens + TOKEN_METER.completion_tokens,
            "hits": TOKEN_METER.cache_hits}


def full_run(run_id):
    t0 = time.time()
    agent = TravelAgent(GOAL, run_id=run_id)
    path = agent.run()
    return time.time() - t0, path


def micro_tools():
    """同一组工具调用：串行 vs 并行（force_refresh 绕过缓存，体现真实 IO 并行收益）。"""
    import tools.weather  # noqa: F401
    import tools.poi
    from tools.base import get_tool

    params_w = {"destination": "苏州", "start_date": "2026-09-10", "days": 2, "force_refresh": True}
    params_p = {"destination": "苏州", "days": 2, "preferences": [], "goal": GOAL}

    t0 = time.time()
    get_tool("weather").run(dict(params_w))
    get_tool("poi").run(dict(params_p))
    serial = time.time() - t0

    t0 = time.time()
    with ThreadPoolExecutor(max_workers=2) as pool:
        pool.submit(get_tool("weather").run, dict(params_w))
        pool.submit(get_tool("poi").run, dict(params_p))
    parallel = time.time() - t0
    return serial, parallel


def main():
    print("=" * 62)
    print(" 性能与 token 基准 · %s" % GOAL)
    print("=" * 62)

    invalidate("weather")
    invalidate("poi")
    before = meter_snapshot()
    cold, path = full_run("bench_cold")
    cold_m = {k: meter_snapshot()[k] - before[k] for k in before}

    before = meter_snapshot()
    warm, _ = full_run("bench_warm")
    warm_m = {k: meter_snapshot()[k] - before[k] for k in before}

    serial, parallel = micro_tools()

    print("""
[1] 全链路端到端
    冷启动（含网络+缓存写入） : %6.2f s   LLM调用 %d 次 / token %d / 缓存命中 %d
    缓存命中（0网络0token）   : %6.2f s   LLM调用 %d 次 / token %d / 缓存命中 %d
    端到端加速比              : %6.1f x

[2] 工具层串行 vs 并行（真实IO）
    串行 weather→poi          : %6.2f s
    并行 weather∥poi          : %6.2f s
    工具层加速比              : %6.1f x

[3] 结论
    - 缓存命中后全链路毫秒级完成，token 消耗为 0（规则模板+知识库检索）
    - 无依赖子任务并行显著缩短网络等待时间
""" % (cold, cold_m["calls"], cold_m["tokens"], cold_m["hits"],
       warm, warm_m["calls"], warm_m["tokens"], warm_m["hits"],
       cold / max(warm, 1e-6), serial, parallel, serial / max(parallel, 1e-6)))
    if path:
        print("方案产物：%s" % path)


if __name__ == "__main__":
    main()

File: kevin/travel-agent/src/main/java/com/travel/agent/Bench.java
package com.travel.agent;

import com.travel.agent.core.TokenMeter;
import com.travel.agent.core.TravelAgent;
import com.travel.agent.tools.ToolRegistry;

import java.util.Map;

/**
 * 性能与 token 基准：冷启动 vs 缓存命中、串行 vs 并行执行对比。java -jar travel-agent.jar bench
 *
 * 输出三组数据，用于验收时量化"加速响应 / 减少 token"两项优化：
 *   1. 冷启动全链路耗时（含真实天气 API + 知识库检索与缓存写入）
 *   2. 缓存命中全链路耗时（0 网络、0 token）
 *   3. 工具串行 vs 并行微基准（DAG 并行加速的依据）
 */
public final class Bench {

    private static final String GOAL = "从上海出发去苏州2天，2人，预算3000元";

    private Bench() {
    }

    private record Snapshot(int calls, long tokens, long hits) {
        static Snapshot now() {
            return new Snapshot(TokenMeter.INSTANCE.getCalls(),
                    TokenMeter.INSTANCE.getTokens(), TokenMeter.INSTANCE.getCacheHits());
        }
    }

    private record RunResult(double elapsed, String path) {
    }

    private static RunResult fullRun(String runId) {
        long t0 = System.nanoTime();
        TravelAgent agent = new TravelAgent(GOAL, runId);
        String path = agent.run();
        return new RunResult((System.nanoTime() - t0) / 1e9, path);
    }

    private static double[] microTools() throws Exception {
        // 同一组工具调用：串行 vs 并行（force_refresh 绕过缓存，体现真实 IO 并行收益）
        Map<String, Object> w = com.travel.agent.util.Json.obj(
                "destination", "苏州", "start_date", "2026-09-10", "days", 2, "force_refresh", true);
        Map<String, Object> p = com.travel.agent.util.Json.obj(
                "destination", "苏州", "days", 2, "preferences", java.util.List.of(), "goal", GOAL);

        long t0 = System.nanoTime();
        ToolRegistry.get("weather").run(new java.util.HashMap<>(w));
        ToolRegistry.get("poi").run(new java.util.HashMap<>(p));
        double serial = (System.nanoTime() - t0) / 1e9;

        t0 = System.nanoTime();
        java.util.concurrent.ExecutorService pool = java.util.concurrent.Executors.newFixedThreadPool(2);
        java.util.List<java.util.concurrent.Future<?>> fs = java.util.List.of(
                pool.submit(() -> {
                    try {
                        ToolRegistry.get("weather").run(new java.util.HashMap<>(w));
                    } catch (Exception ignored) {
                    }
                }),
                pool.submit(() -> {
                    try {
                        ToolRegistry.get("poi").run(new java.util.HashMap<>(p));
                    } catch (Exception ignored) {
                    }
                }));
        for (java.util.concurrent.Future<?> f : fs) {
            f.get();
        }
        pool.shutdown();
        pool.awaitTermination(1, java.util.concurrent.TimeUnit.HOURS);
        double parallel = (System.nanoTime() - t0) / 1e9;
        return new double[]{serial, parallel};
    }

    public static void main(String[] args) throws Exception {
        System.out.println("=".repeat(62));
        System.out.println(" 性能与 token 基准 · " + GOAL);
        System.out.println("=".repeat(62));

        ToolRegistry.invalidate("weather");
        ToolRegistry.invalidate("poi");
        Snapshot before = Snapshot.now();
        RunResult cold = fullRun("bench_cold");
        Snapshot afterCold = Snapshot.now();
        int coldCalls = afterCold.calls() - before.calls();
        long coldTokens = afterCold.tokens() - before.tokens();
        long coldHits = afterCold.hits() - before.hits();

        before = Snapshot.now();
        RunResult warm = fullRun("bench_warm");
        Snapshot afterWarm = Snapshot.now();
        int warmCalls = afterWarm.calls() - before.calls();
        long warmTokens = afterWarm.tokens() - before.tokens();
        long warmHits = afterWarm.hits() - before.hits();

        double[] sp = microTools();

        System.out.printf("""
                %n[1] 全链路端到端
                    冷启动（含网络+缓存写入） : %6.2f s   LLM调用 %d 次 / token %d / 缓存命中 %d
                    缓存命中（0网络0token）   : %6.2f s   LLM调用 %d 次 / token %d / 缓存命中 %d
                    端到端加速比              : %6.1f x

                [2] 工具层串行 vs 并行（真实IO）
                    串行 weather→poi          : %6.2f s
                    并行 weather∥poi          : %6.2f s
                    工具层加速比              : %6.1f x

                [3] 结论
                    - 缓存命中后全链路毫秒级完成，token 消耗为 0（规则模板+知识库检索）
                    - 无依赖子任务并行显著缩短网络等待时间
                %n""", cold.elapsed(), coldCalls, coldTokens, coldHits,
                warm.elapsed(), warmCalls, warmTokens, warmHits,
                cold.elapsed() / Math.max(warm.elapsed(), 1e-6),
                sp[0], sp[1], sp[0] / Math.max(sp[1], 1e-6));
        if (cold.path() != null) {
            System.out.println("方案产物：" + cold.path());
        }
    }
}
